File: 03_use_cases/13_macos_cua/src/metacua/Commands.swift
import CoreGraphics
import Foundation

/// Common setup for every action: verify permission, then bring the app forward.
private func preflight(app: String) throws {
  try requireTrust()
  try activateApp(app)
}

func runClick(_ raw: [String]) throws {
  let args = try Args(raw)
  let x = try args.requiredDouble("x")
  let y = try args.requiredDouble("y")
  let clickCount = try args.int("click-count") ?? args.int("clicks") ?? 1
  let button = try MouseButton.parse(args.string(["mouse-button", "button"]) ?? "left")
  let app = try args.requiredString("app")
  let point = CGPoint(x: x, y: y)

  try preflight(app: app)
  performClick(at: point, button: button, clickCount: clickCount, source: eventSource())
  print("clicked \(button.rawValue) x\(max(1, clickCount)) at (\(x), \(y)) in \(app)")
}

func runShot(_ raw: [String]) throws {
  let args = try Args(raw)
  try requireScreenRecording()
  let scale = try args.double("scale") ?? 1.0
  let shot = try captureScreenshot(scale: scale)
  guard let data = Data(base64Encoded: shot.pngBase64) else {
    throw CLIError("failed to decode the captured screenshot")
  }
  let outPath =
    args.string(["out", "o"])
    ?? FileManager.default.temporaryDirectory.appendingPathComponent("metacua-shot.png").path
  let url = URL(fileURLWithPath: (outPath as NSString).expandingTildeInPath)
  try data.write(to: url)
  let sizeLabel =
    shot.imageWidth == shot.width && shot.imageHeight == shot.height
    ? "\(shot.imageWidth)x\(shot.imageHeight)"
    : "\(shot.imageWidth)x\(shot.imageHeight) image, \(shot.width)x\(shot.height) coords"
  print("captured \(sizeLabel) screenshot -> \(url.path) (\(data.count) bytes)")
}

func runMove(_ raw: [String]) throws {
  let args = try Args(raw)
  let x = try args.requiredDouble("x")
  let y = try args.requiredDouble("y")
  let app = try args.requiredString("app")
  let point = CGPoint(x: x, y: y)

  try preflight(app: app)
  moveMouse(to: point, source: eventSource())
  print("moved pointer to (\(x), \(y)) in \(app)")
}

func runScroll(_ raw: [String]) throws {
  let args = try Args(raw)
  let direction = try ScrollDirection.parse(try args.requiredString("direction"))
  let pages = try args.int("pages") ?? 1
  let linesPerPage = try args.int("lines-per-page") ?? 10
  let app = try args.requiredString("app")

  let xOpt = try args.double("x")
  let yOpt = try args.double("y")
  if (xOpt == nil) != (yOpt == nil) {
    throw CLIError("--x and --y must be provided together", code: 2)
  }
  var point: CGPoint?
  if let x = xOpt, let y = yOpt {
    point = CGPoint(x: x, y: y)
  }

  try preflight(app: app)
  performScroll(
    direction: direction, pages: pages, linesPerPage: linesPerPage, at: point, source: eventSource()
  )
  print("scrolled \(direction.rawValue) \(pages) page(s) in \(app)")
}

func runDrag(_ raw: [String]) throws {
  let args = try Args(raw)
  let fromX = try args.requiredDouble("from-x")
  let fromY = try args.requiredDouble("from-y")
  let toX = try args.requiredDouble("to-x")
  let toY = try args.requiredDouble("to-y")
  let button = try MouseButton.parse(args.string(["mouse-button", "button"]) ?? "left")
  let app = try args.requiredString("app")
  let from = CGPoint(x: fromX, y: fromY)
  let to = CGPoint(x: toX, y: toY)

  try preflight(app: app)
  performDrag(from: from, to: to, button: button, source: eventSource())
  print("dragged from (\(fromX), \(fromY)) to (\(toX), \(toY)) in \(app)")
}

func runPressKey(_ raw: [String]) throws {
  let args = try Args(raw)
  let key = try args.requiredString("key")
  let app = try args.requiredString("app")
  let combo = try parseKeyCombo(key)

  try preflight(app: app)
  performKeyCombo(code: combo.code, flags: combo.flags, source: eventSource())
  print("pressed '\(key)' in \(app)")
}

func runTypeText(_ raw: [String]) throws {
  let args = try Args(raw)
  let text = try args.requiredString("text")
  let app = try args.requiredString("app")

  try preflight(app: app)
  try performTypeText(text, source: eventSource())
  print("inserted \(text.count) character(s) in \(app)")
}

File: 03_use_cases/13_macos_cua/python/metacua/commands.py
"""Manual tool subcommands: click, moveto, scroll, drag, press-key, type-text, shot."""

import base64
import os

from Quartz import CGPointMake

from .app_control import activate_app
from .args import Args
from .errors import CLIError
from .mouse import (
    MouseButton,
    ScrollDirection,
    event_source,
    move_mouse,
    perform_click,
    perform_drag,
    perform_scroll,
)
from .keyboard import parse_key_combo, perform_key_combo, perform_type_text
from .permissions import require_screen_recording, require_trust
from .screenshot import capture_screenshot


def _preflight(app: str) -> None:
    """Common setup for every action: verify permission, then bring the app forward."""
    require_trust()
    activate_app(app)


def run_click(raw) -> None:
    args = Args(raw)
    x = args.required_double("x")
    y = args.required_double("y")
    click_count = args.int("click-count") or args.int("clicks") or 1
    button = MouseButton.parse(args.string(["mouse-button", "button"]) or "left")
    app = args.required_string("app")
    point = CGPointMake(x, y)

    _preflight(app)
    perform_click(point, button, click_count, event_source())
    print(f"clicked {button.value} x{max(1, click_count)} at ({x}, {y}) in {app}")


def run_shot(raw) -> None:
    args = Args(raw)
    require_screen_recording()
    scale = args.double("scale")
    scale = 1.0 if scale is None else scale
    shot = capture_screenshot(scale)
    try:
        data = base64.b64decode(shot.png_base64)
    except (ValueError, TypeError):
        raise CLIError("failed to decode the captured screenshot")
    out_path = args.string(["out", "o"]) or os.path.join(
        os.environ.get("TMPDIR", "/tmp"), "metacua-shot.png"
    )
    path = os.path.expanduser(out_path)
    with open(path, "wb") as handle:
        handle.write(data)
    if shot.image_width == shot.width and shot.image_height == shot.height:
        size_label = f"{shot.image_width}x{shot.image_height}"
    else:
        size_label = (
            f"{shot.image_width}x{shot.image_height} image, {shot.width}x{shot.height} coords"
        )
    print(f"captured {size_label} screenshot -> {path} ({len(data)} bytes)")


def run_move(raw) -> None:
    args = Args(raw)
    x = args.required_double("x")
    y = args.required_double("y")
    app = args.required_string("app")
    point = CGPointMake(x, y)

    _preflight(app)
    move_mouse(point, event_source())
    print(f"moved pointer to ({x}, {y}) in {app}")


def run_scroll(raw) -> None:
    args = Args(raw)
    direction = ScrollDirection.parse(args.required_string("direction"))
    pages = args.int("pages") or 1
    lines_per_page = args.int("lines-per-page") or 10
    app = args.required_string("app")

    x_opt = args.double("x")
    y_opt = args.double("y")
    if (x_opt is None) != (y_opt is None):
        raise CLIError("--x and --y must be provided together", code=2)
    point = None
    if x_opt is not None and y_opt is not None:
        point = CGPointMake(x_opt, y_opt)

    _preflight(app)
    perform_scroll(direction, pages, lines_per_page, point, event_source())
    print(f"scrolled {direction.value} {pages} page(s) in {app}")


def run_drag(raw) -> None:
    args = Args(raw)
    from_x = args.required_double("from-x")
    from_y = args.required_double("from-y")
    to_x = args.required_double("to-x")
    to_y = args.required_double("to-y")
    button = MouseButton.parse(args.string(["mouse-button", "button"]) or "left")
    app = args.required_string("app")
    from_point = CGPointMake(from_x, from_y)
    to_point = CGPointMake(to_x, to_y)

    _preflight(app)
    perform_drag(from_point, to_point, button, event_source())
    print(f"dragged from ({from_x}, {from_y}) to ({to_x}, {to_y}) in {app}")


def run_press_key(raw) -> None:
    args = Args(raw)
    key = args.required_string("key")
    app = args.required_string("app")
    code, flags = parse_key_combo(key)

    _preflight(app)
    perform_key_combo(code, flags, event_source())
    print(f"pressed '{key}' in {app}")


def run_type_text(raw) -> None:
    args = Args(raw)
    text = args.required_string("text")
    app = args.required_string("app")

    _preflight(app)
    perform_type_text(text, event_source())
    print(f"typed {len(text)} character(s) in {app}")
